spec/src/consensus.rs

```diff
@@ -340,82 +377,105 @@ impl Consensus {
-
-        let epoch_ext = if last_uncles_count > 0 {
-            let last_epoch_duration = header
+        let last_epoch_duration = cmp::max(
+            header
                 .timestamp()
-                .saturating_sub(last_block_header_in_previous_epoch.timestamp());
-            if last_epoch_duration == 0 {
-                return None;
-            }
+                .saturating_sub(last_block_header_in_previous_epoch.timestamp())
+                / MILLISECOND_TO_SECOND,
+            1,
+        );
+
+        let last_epoch_hash_rate =
+            last_difficulty * (last_epoch_length + last_uncles_count) / last_epoch_duration;
+
+        let adjusted_last_epoch_hash_rate = self.bounding_hash_rate(
+            last_epoch_hash_rate,
+            last_epoch.previous_epoch_hash_rate().to_owned(),
+        );
+
+        debug_assert!(
+            adjusted_last_epoch_hash_rate > U256::zero(),
+            "adjusted_last_epoch_hash_rate should greater than one"
+        );
+
+        // (2) Computing the Next Epoch’s Main Chain Block Number
+        let target_recip = self.orphan_rate_target_recip();
+        let epoch_duration_target = self.epoch_duration_target();
 
+        let (next_epoch_length, bound) = if last_uncles_count == 0 {
+            (self.bounding_epoch_length(last_epoch_length * TAU).0, true)
+        } else {
             let numerator =
                 (last_uncles_count + last_epoch_length) * epoch_duration_target * last_epoch_length;
             let denominator = (target_recip + 1) * last_uncles_count * last_epoch_duration;
             let raw_next_epoch_length = numerator / denominator;
-            let next_epoch_length = self.revision_epoch_length(raw_next_epoch_length);
-
-            let raw_difficulty =
-                last_difficulty * U256::from(last_uncles_count) * U256::from(target_recip)
-                    / U256::from(last_epoch_length);
-
-            let difficulty =
-                self.revision_epoch_difficulty(last_difficulty.clone(), raw_difficulty);
-
-            let block_reward = Capacity::shannons(self.epoch_reward().as_u64() / next_epoch_length);
-            let remainder_reward =
-                Capacity::shannons(self.epoch_reward().as_u64() % next_epoch_length);
-
-            EpochExt::new(
-                last_epoch.number() + 1,     // number
-                block_reward,
-                remainder_reward,            // remainder_reward
-                header.hash().to_owned(),    // last_block_hash_in_previous_epoch
-                header.number() + 1,         // start
-                next_epoch_length,           // length
-                difficulty                   // difficulty,
-            )
+            self.bounding_epoch_length(raw_next_epoch_length)
+        };
+
+        // (3) Determining the Next Epoch’s Difficulty
+        let diff_numerator = &adjusted_last_epoch_hash_rate * epoch_duration_target;
+        let next_epoch_diff = if bound {
```

Comment:
Using the single unified equation is better than calculate differently on different cases.